File: testers/duckdb/java/src/java/testdb/DuckDbTestHelper.java
package testdb;

import dev.typr.foundations.Connection;
import dev.typr.foundations.ConnectionRead;
import dev.typr.foundations.SqlConsumer;
import dev.typr.foundations.SqlFunction;
import dev.typr.foundations.Transactor;
import dev.typr.foundations.connect.ConnectionSettings;
import dev.typr.foundations.connect.DuckDbConfig;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class DuckDbTestHelper {
  private static final DuckDbConfig CONFIG = DuckDbConfig.inMemory().build();

  private static String readSchema() {
    try {
      return Files.readString(Path.of("sql-init/duckdb/00-schema.sql"));
    } catch (IOException e) {
      throw new RuntimeException("Failed to read DuckDB schema", e);
    }
  }

  private static final ConnectionSettings SETTINGS =
      ConnectionSettings.builder().connectionInitSql(readSchema()).build();

  private static final Transactor TRANSACTOR = Transactor.create(CONFIG, SETTINGS).rollbackOnly();

  public static <T> T apply(SqlFunction<Connection, T> f) {
    return TRANSACTOR.transact(f);
  }

  public static void run(SqlConsumer<Connection> f) {
    TRANSACTOR.transactVoid(f);
  }

  public static <T> T applyRead(SqlFunction<ConnectionRead, T> f) {
    return TRANSACTOR.transactRead(f);
  }
}

File: testers/duckdb/kotlin/src/kotlin/testdb/DuckDbTestHelper.kt
package testdb

import dev.typr.foundationskt.Connection
import dev.typr.foundationskt.ConnectionRead
import dev.typr.foundationskt.Transactor
import dev.typr.foundationskt.connect.ConnectionSettings
import dev.typr.foundationskt.connect.DuckDbConfig
import java.nio.file.Files
import java.nio.file.Path

object DuckDbTestHelper {
    private val CONFIG: DuckDbConfig = DuckDbConfig.inMemory().build()
    private val SETTINGS: ConnectionSettings =
        ConnectionSettings.builder()
            .connectionInitSql(Files.readString(Path.of("sql-init/duckdb/00-schema.sql")))
            .build()

    private val TRANSACTOR: Transactor = Transactor.create(CONFIG, SETTINGS).rollbackOnly()

    fun <T> apply(f: (Connection) -> T): T = TRANSACTOR.transact(f)

    fun run(f: (Connection) -> Unit) = TRANSACTOR.transact { conn -> f(conn) }

    fun <T> applyRead(f: (ConnectionRead) -> T): T = TRANSACTOR.transactRead(f)
}
